# Mini Project: Transfer Learning with Keras

Transfer learning is a machine learning technique where a model trained on one task is used as a starting point to solve a different but related task. Instead of training a model from scratch, transfer learning leverages the knowledge learned from the source task and applies it to the target task. This approach is especially useful when the target task has limited data or computational resources.

In transfer learning, the pre-trained model, also known as the "base model" or "source model," is typically trained on a large dataset and a more general problem (e.g., image classification on ImageNet, a vast dataset with millions of labeled images). The knowledge learned by the base model in the form of feature representations and weights captures common patterns and features in the data.

To perform transfer learning, the following steps are commonly followed:

1. Pre-training: The base model is trained on a source task using a large dataset, which can take a considerable amount of time and computational resources.

2. Feature Extraction: After pre-training, the base model is used as a feature extractor. The last few layers (classifier layers) of the model are discarded, and the remaining layers (feature extraction layers) are retained. These layers serve as feature extractors, producing meaningful representations of the data.

3. Fine-tuning: The feature extraction layers and sometimes some of the earlier layers are connected to a new set of layers, often called the "classifier layers" or "task-specific layers." These layers are randomly initialized, and the model is trained on the target task with a smaller dataset. The weights of the base model can be frozen during fine-tuning, or they can be allowed to be updated with a lower learning rate to fine-tune the model for the target task.

Transfer learning has several benefits:

1. Reduced training time and resource requirements: Since the base model has already learned generic features, transfer learning can save time and resources compared to training a model from scratch.

2. Improved generalization: Transfer learning helps the model generalize better to the target task, especially when the target dataset is small and dissimilar from the source dataset.

3. Better performance: By starting from a model that is already trained on a large dataset, transfer learning can lead to better performance on the target task, especially in scenarios with limited data.

4. Effective feature extraction: The feature extraction layers of the pre-trained model can serve as powerful feature extractors for different tasks, even when the task domains differ.

Transfer learning is commonly used in various domains, including computer vision, natural language processing (NLP), and speech recognition, where pre-trained models are fine-tuned for specific applications like object detection, sentiment analysis, or speech-to-text.

In this mini-project you will perform fine-tuning using Keras with a pre-trained VGG16 model on the CIFAR-10 dataset.

First, import all the libraries you'll need.

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.applications import VGG16
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, GlobalAveragePooling2D
from tensorflow.keras.optimizers import Adam, AdamW
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split

The CIFAR-10 dataset is a widely used benchmark dataset in the field of computer vision and machine learning. It stands for the "Canadian Institute for Advanced Research 10" dataset. CIFAR-10 was created by researchers at the CIFAR institute and was originally introduced as part of the Neural Information Processing Systems (NIPS) 2009 competition.

The dataset consists of 60,000 color images, each of size 32x32 pixels, belonging to ten different classes. Each class contains 6,000 images. The ten classes in CIFAR-10 are:

1. Airplane
2. Automobile
3. Bird
4. Cat
5. Deer
6. Dog
7. Frog
8. Horse
9. Ship
10. Truck

The images are evenly distributed across the classes, making CIFAR-10 a balanced dataset. The dataset is divided into two sets: a training set and a test set. The training set contains 50,000 images, while the test set contains the remaining 10,000 images.

CIFAR-10 is often used for tasks such as image classification, object recognition, and transfer learning experiments. The relatively small size of the images and the variety of classes make it a challenging dataset for training machine learning models, especially deep neural networks. It also serves as a good dataset for teaching and learning purposes due to its manageable size and straightforward class labels.

Here are your tasks:

1. Load the CIFAR-10 dataset after referencing the documentation [here](https://keras.io/api/datasets/cifar10/).
2. Normalize the pixel values so they're all in the range [0, 1].
3. Apply One Hot Encoding to the train and test labels using the [to_categorical](https://www.tensorflow.org/api_docs/python/tf/keras/utils/to_categorical) function.
4. Further split the the training data into training and validation sets using [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html). Use only 10% of the data for validation.  

In [ ]:
# Load the CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

170498071/170498071 [==============================] - 2s 0us/step


In [ ]:
print("Training Size:",x_train.shape)
print("Test Size:", x_test.shape)
print("Training Label Size:", y_train.shape)
print("Test Label Size:", y_test.shape)

Training Size: (50000, 32, 32, 3)
Test Size: (10000, 32, 32, 3)
Training Label Size: (50000, 1)
Test Label Size: (10000, 1)


In [ ]:
# Normalize the pixel values to [0, 1]
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

In [ ]:
# One-hot encode the labels
y_train = to_categorical(y_train, num_classes=10)
y_test = to_categorical(y_test, num_classes=10)

In [ ]:
# Split the data into training and validation sets
x_train2, x_test2, y_train2, y_test2 = train_test_split(x_train,
                                                        y_train,
                                                        test_size=0.1,
                                                        random_state=42)

VGG16 (Visual Geometry Group 16) is a deep convolutional neural network architecture that was developed by the Visual Geometry Group at the University of Oxford. It was proposed by researchers Karen Simonyan and Andrew Zisserman in their paper titled "Very Deep Convolutional Networks for Large-Scale Image Recognition," which was presented at the International Conference on Learning Representations (ICLR) in 2015.

The VGG16 architecture gained significant popularity for its simplicity and effectiveness in image classification tasks. It was one of the pioneering models that demonstrated the power of deeper neural networks for visual recognition tasks.

Key characteristics of the VGG16 architecture:

1. Architecture: VGG16 consists of a total of 16 layers, hence the name "16." These layers are stacked one after another, forming a deep neural network.

2. Convolutional Layers: The main building blocks of VGG16 are the convolutional layers. It primarily uses 3x3 convolutional filters throughout the network, which allows it to capture local features effectively.

3. Max Pooling: After each set of convolutional layers, VGG16 applies max-pooling layers with 2x2 filters and stride 2, which halves the spatial dimensions (width and height) of the feature maps and reduces the number of parameters.

4. Fully Connected Layers: Towards the end of the network, VGG16 has fully connected layers that act as a classifier to make predictions based on the learned features.

5. Activation Function: The network uses the Rectified Linear Unit (ReLU) activation function for all hidden layers, which helps with faster convergence during training.

6. Number of Filters: The number of filters in each convolutional layer is relatively small compared to more recent architectures like ResNet or InceptionNet. However, stacking multiple layers allows VGG16 to learn complex hierarchical features.

7. Output Layer: The output layer consists of 1000 units, corresponding to 1000 ImageNet classes. VGG16 was originally trained on the large-scale ImageNet dataset, which contains millions of images from 1000 different classes.

VGG16 was instrumental in showing that increasing the depth of a neural network can significantly improve its performance on image recognition tasks. However, the main drawback of VGG16 is its high number of parameters, making it computationally expensive and memory-intensive to train. Despite this limitation, VGG16 remains an essential benchmark architecture and has paved the way for even deeper and more efficient models in the field of computer vision, such as ResNet, DenseNet, and EfficientNet.

Here are your tasks:

1. Load [VGG16](https://keras.io/api/applications/vgg/#vgg16-function) as a base model. Make sure to exclude the top layer.
2. Freeze all the layers in the base model. We'll be using these weights as a feature extraction layer to forward to layers that are trainable.

In [ ]:
# Load the pre-trained VGG16 model (excluding the top classifier)
vgg16_base = VGG16(weights='imagenet',
                   include_top=False,
                   input_shape=(32, 32, 3))

58889256/58889256 [==============================] - 0s 0us/step


In [ ]:
# Freeze the layers in the base model
for layer in vgg16_base.layers:
    layer.trainable = False

Now, we'll add some trainable layers to the base model.

1. Using the base model, add a [GlobalAveragePooling2D](https://keras.io/api/layers/pooling_layers/global_average_pooling2d/) layer, followed by a [Dense](https://keras.io/api/layers/core_layers/dense/) layer of length 256 with ReLU activation. Finally, add a classification layer with 10 units, corresponding to the 10 CIFAR-10 classes, with softmax activation.
2. Create a Keras [Model](https://keras.io/api/models/model/) that takes in approproate inputs and outputs.

In [ ]:
# Add a global average pooling layer
x = GlobalAveragePooling2D()(vgg16_base.output)

In [ ]:
print(x.shape)

(None, 512)


In [ ]:
# Add a fully connected layer with 256 units and ReLU activation
dense256 = Dense(256, activation='relu')(x)

In [ ]:
print(dense256.shape)

(None, 256)


In [ ]:
# Add the final classification layer with 10 units (for CIFAR-10 classes) and softmax activation
predictions = Dense(10, activation='softmax')(dense256)

In [ ]:
print(predictions.shape)

(None, 10)


In [ ]:
# Create the fine-tuned model
ftmodel = Model(inputs=vgg16_base.input, outputs=predictions)

In [ ]:
print(ftmodel.summary())

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 32, 32, 3)]       0         
                                                                 
 block1_conv1 (Conv2D)       (None, 32, 32, 64)        1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 32, 32, 64)        36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 16, 16, 64)        0         
                                                                 
 block2_conv1 (Conv2D)       (None, 16, 16, 128)       73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 16, 16, 128)       147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 8, 8, 128)         0     

With your model complete it's time to train it and assess its performance.

1. Compile your model using an appropriate loss function. Feel free to play around with the optimizer, but a good starting optimizer might be Adam with a learning rate of 0.001.
2. Fit your model on the training data. Use the validation data to print the accuracy for each epoch. Try training for 10 epochs. Note, training can take a few hours so go ahead and grab a cup of coffee.

**Optional**: See if you can implement an [Early Stopping](https://keras.io/api/callbacks/early_stopping/) criteria as a callback function.

In [ ]:
# Compile the model
ftmodel.compile(optimizer=Adam(learning_rate=0.001),
                loss='categorical_crossentropy',
                metrics=['accuracy'])

In [ ]:
# Train the model
ftmodel.fit(x_train2, y_train2,
            batch_size=32,
            epochs=10,
            validation_data=(x_test2, y_test2))

Epoch 1/10
1407/1407 [==============================] - 56s 39ms/step - loss: 1.3602 - accuracy: 0.5262 - val_loss: 1.2188 - val_accuracy: 0.5692
Epoch 2/10
1407/1407 [==============================] - 54s 38ms/step - loss: 1.1797 - accuracy: 0.5866 - val_loss: 1.1820 - val_accuracy: 0.5820
Epoch 3/10
1407/1407 [==============================] - 54s 38ms/step - loss: 1.1139 - accuracy: 0.6101 - val_loss: 1.1315 - val_accuracy: 0.5996
Epoch 4/10
1407/1407 [==============================] - 54s 38ms/step - loss: 1.0637 - accuracy: 0.6269 - val_loss: 1.1517 - val_accuracy: 0.5968
Epoch 5/10
1407/1407 [==============================] - 54s 38ms/step - loss: 1.0212 - accuracy: 0.6413 - val_loss: 1.1261 - val_accuracy: 0.6156
Epoch 6/10
1407/1407 [==============================] - 54s 39ms/step - loss: 0.9821 - accuracy: 0.6566 - val_loss: 1.1100 - val_accuracy: 0.6138
Epoch 7/10
1407/1407 [==============================] - 54s 39ms/step - loss: 0.9459 - accuracy: 0.6695 - val_loss: 1.1015 -

With your model trained, it's time to assess how well it performs on the test data.

1. Use your trained model to calculate the accuracy on the test set. Is the model performance better than random?
2. Experiment! See if you can tweak your model to improve performance.  

In [ ]:
# Evaluate the model on the test set
test_loss, test_accuracy = ftmodel.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy * 100:.2f}%")

313/313 [==============================] - 11s 34ms/step - loss: 1.1312 - accuracy: 0.6129
Test Accuracy: 61.29%




> It's interesting the accuracy is not that great at about 61.3%. The training accuracy was way better. Perhaps the model was overfitted using 90% of the data for training?



In [ ]:
# Attempt 2 using optimizer=AdamW
# Compile the model
ftmodel.compile(optimizer=AdamW(learning_rate=0.001),
                loss='categorical_crossentropy',
                metrics=['accuracy'])

In [ ]:
# Train the model (again)
ftmodel.fit(x_train2, y_train2,
            batch_size=32,
            epochs=10,
            validation_data=(x_test2, y_test2))

Epoch 1/10
1407/1407 [==============================] - 54s 38ms/step - loss: 0.8238 - accuracy: 0.7110 - val_loss: 1.1125 - val_accuracy: 0.6158
Epoch 2/10
1407/1407 [==============================] - 54s 38ms/step - loss: 0.7984 - accuracy: 0.7200 - val_loss: 1.1092 - val_accuracy: 0.6238
Epoch 3/10
1407/1407 [==============================] - 54s 38ms/step - loss: 0.7723 - accuracy: 0.7282 - val_loss: 1.1185 - val_accuracy: 0.6274
Epoch 4/10
1407/1407 [==============================] - 53s 38ms/step - loss: 0.7499 - accuracy: 0.7372 - val_loss: 1.1381 - val_accuracy: 0.6248
Epoch 5/10
1407/1407 [==============================] - 54s 38ms/step - loss: 0.7253 - accuracy: 0.7462 - val_loss: 1.1222 - val_accuracy: 0.6222
Epoch 6/10
1407/1407 [==============================] - 54s 38ms/step - loss: 0.7025 - accuracy: 0.7543 - val_loss: 1.1237 - val_accuracy: 0.6228
Epoch 7/10
1407/1407 [==============================] - 54s 38ms/step - loss: 0.6832 - accuracy: 0.7620 - val_loss: 1.1714 -

In [ ]:
# Evaluate the model on the test set
test_loss, test_accuracy = ftmodel.evaluate(x_test2, y_test2)
print(f"Test Accuracy: {test_accuracy * 100:.2f}%")

157/157 [==============================] - 5s 34ms/step - loss: 1.2163 - accuracy: 0.6200
Test Accuracy: 62.00%




> Training using the AdamW optimizer did not improve the accuracy that much over using Adam.

> Over 10 epochs there didn't appear an opportunity to use EarlyStopping as the accuracy continued to improve. I think I could have tried 20 or 30 epochs but would need more time for that.

